### PR TITLE
Query Execution Plan Expensive Operator Highlighting

### LINKED ISSUE #20260: Highlight the most expensive operators in a plan
Identify the most expensive operator(s) in a plan with a color notation (or something else if more accessible) to help users with prioritizing what to investigate in a plan.  Historically, cost has been used to identify "what is expensive", but other options include Actual Elapsed CPU Time (ms) or Actual Elapsed Time (ms).  Theoretically, folks could also be interested in operators with the highest number of rows (estimated or actual), Estimated I/O Cost, or Estimated CPU Cost.

### PR DESCRIPTION
**Pull request by lewis-sanchez**
This PR fixes #20260.

This PR adds a new action bar icon to query execution plans and allows a user to locate a node/operation based on the following metrics:
image

This new widget can be opened by clicking on the following button in the action bar: 
image

*** By default the cost metric is chosen, but the user can switch between any of the other options seamlessly.

Clicking on Find results in node's being enclosed in an orange outlined box like this:
image

In the event that a metric isn't present in a graph like attempting to look for "Actual Elapsed Time" in an estimated plan, than an information popup will appear stating that the property wasn't found.
image

Below is a gif that demonstrates how the widget works:
Highlight Expensive Operator

### CODE REVIEW

**Comment by aasimkhan30** [comment]
Let's have a discussion about the design of this widget. The button says find which is not what it does. Also, it is not clear what the widget is supposed to do.

**Review by erinstellato-ms — COMMENTED** [review]
Ok, I think the button needs to be named something different than "Find".  I realize we might have limited space for text but we need to indicate that we are finding the most expensive operator.

Also, I'm not clear how trying to find most elapsed time would not result in any operator being selected.

**Comment by alanrenmsft** [comment]
the dropdown and the Find button is overlapping. please fix.

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+			}
+		};
+
+		const findButton = new Button(this.container, {
+			title: localize('findExpensiveOperationButtonTitle', 'Find Expensive Operation (Enter)')
+		});
+		findButton.setWidth('60px');
+		findButton.label = localize('findExpensiveOperationApplyButton', 'Find');
+
+		findButton.onDidClick(async e => {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:103]: The widget should be disposable and all disposable objects it creates should be registered. 

(I'd probably make the base widget itself extend Disposable - it looks like you aren't doing this for any of the other widgets either)

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+		super(FindExpensiveOperationAction.ID, FindExpensiveOperationAction.LABEL, searchIconClassNames);
+	}
+
+	public override async run(context: FindExpensiveOperationWidget): Promise<void> {
+		const expensiveOperationDelegate: (cell: AzDataGraphCell) => number | undefined = context.getExpensiveOperationDelegate();
+
+		context.executionPlanDiagram.clearExpensiveOperatorHighlighting();
+		let result = context.executionPlanDiagram.highlightExpensiveOperator(expensiveOperationDelegate);
+		if (!result) {
+			context.notificationService.info(localize('invalidPropertyExecutionPlanMetric', 'Unable to locate a node using the specified metric.'));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:188]: I would suggest saying what the metric was too - and maybe reword it to be something like `No nodes found with the '${MetricName}' metric.`

I'm also not a huge fan of the toast notifications being used for this. Taking the find widget as inspiration it would be better to have the widget itself indicate that nothing was found.

**Review by Charles-Gagnon — COMMENTED** [review]
Was there a design discussion for this? I think it would be good to make a doc for this and go over it with some folks to make sure this is easy and intuitive to use and solves whatever problem we're looking to have it address. 

A couple of general thoughts : 

1. Why does this only find the most expensive operator? What if I want to see the 2nd or 3rd most expensive - is there a way to do that easily?
2. Is this bar shown all the time? It might be better to look into having a keybinding bring it up (similar to how the find widget works)
3. If we know that some values won't be present in certain circumstances (like "actual" values in estimated plans) then it would be nice to just not show them at all.

**Comment by lewis-sanchez** [comment]
> Was there a design discussion for this? I think it would be good to make a doc for this and go over it with some folks to make sure this is easy and intuitive to use and solves whatever problem we're looking to have it address.
> 
> A couple of general thoughts :
> 
> 1. Why does this only find the most expensive operator? What if I want to see the 2nd or 3rd most expensive - is there a way to do that easily?
> 2. Is this bar shown all the time? It might be better to look into having a keybinding bring it up (similar to how the find widget works)
> 3. If we know that some values won't be present in certain circumstances (like "actual" values in estimated plans) then it would be nice to just not show them at all.

Erin and I discussed what we had in mind for the behavior. We wrote a one pager where we discussed making the find expensive operator widget flexible enough to find a node based on a user specified metric instead of defaulting to a metric like cost every time. The suggestion for an adjustable metric came from user feedback and is mentioned in the one pager for this feature.

The answers to questions 1-3:
1. We discussed highlighting several nodes and decided to highlight just a single node.
2. The bar is only shown when the user opens it from the action bar. [(ctrl/cmd + f) is already used by the "Find Node" widget]
3. I definitely like this suggestion and I will look into it.

File: src/sql/azdata.proposed.d.ts [hunk 3]
```diff
@@ -1523,6 +1531,13 @@ declare module 'azdata' {
 			 */
 			displayValue: string | number | boolean;
 		}
+
+		export interface Dictionary {
```
Comment (kburtram) [inline src/sql/azdata.proposed.d.ts:1535]: this seems like a very generic interface to exporting from the azdata extensibility namespace.  I wonder if it's better to just declare the JS dictionary/array where this interface is used, or give this a more scoped name like `ExecutionPlanMetricsDictionary`.  Personally, I think we can use the JS type directly without loss of readability/etc.

File: src/sql/workbench/contrib/executionPlan/browser/azdataGraphView.ts [hunk 4]
```diff
@@ -323,6 +330,31 @@ export class AzdataGraphView {
 		if (node.description) {
 			diagramNode.description = node.description;
 		}
+
+		if (node.cost) {
```
Comment (kburtram) [inline src/sql/workbench/contrib/executionPlan/browser/azdataGraphView.ts:334]: why do we need to null check before assigning?  are we trying to preserve the values in `diagramNode` when the corresponding value is null in `node`?  If not, I think directly assigning all these properties regardless of state is a more compact/efficient/readable style.  If we are trying to preserve values (or there is some cost to assigning values, e.g. expensive property setters, etc.) then this approach makes sense to me.

File: src/sql/workbench/contrib/query/browser/query.contribution.ts [hunk 5]
```diff
@@ -450,6 +450,12 @@ const queryEditorConfiguration: IConfigurationNode = {
 			'type': 'boolean',
 			'default': false,
 			'description': localize('queryEditor.promptToSaveGeneratedFiles', "Prompt to save generated SQL files")
+		},
+		'queryEditor.executionPlan.expensiveOperationMetric': {
+			'type': 'string',
+			'enum': ['off', 'actualElapsedTime', 'actualElapsedCpuTime', 'cost', 'subtreeCost', 'actualNumberOfRowsForAllExecutions', 'numberOfRowsRead'],
```
Comment (kburtram) [inline src/sql/workbench/contrib/query/browser/query.contribution.ts:456]: do also need an `enumDescriptions` array? if these identifiers will be user facing we probably do.

**Review by kburtram — APPROVED** [review]
Overall, looks good to me.  There are a few comments that should probably be addressed before merged (both mine and existing).  But otherwise looks great.

File: src/sql/azdata.proposed.d.ts (hunk 3 above)
Comment (Charles-Gagnon) [reply src/sql/azdata.proposed.d.ts:1535]: +1, we shouldn't be making our own version of stuff that already exists. You can just use `Record<string, string | undefined>` as the type. We probably don't want to use Dictionary since that doesn't transfer well over RPC.

File: src/sql/azdata.proposed.d.ts [hunk 6]
```diff
@@ -1333,6 +1337,10 @@ declare module 'azdata' {
 			 * Cost string for the node
 			 */
 			costDisplayString: string;
+			/**
+			 * Row metrics for the node
+			 */
+			rowMetrics: Dictionary;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts:1343]: please add an example in the comments

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+		this._register(findExpensiveOperationAction);
+
+		const clearExpensiveOperationAction = new TurnOffExpensiveOperationAction();
+		this._register(clearExpensiveOperationAction);
+
+		const cancelExpensiveOperationAction = new CancelExpensiveOperationAction();
+		this._register(cancelExpensiveOperationAction);
+
+		const self = this;
+		this._operationNameSelectBox.selectElem.onkeydown = async (ev) => {
```
Comment (alanrenmsft) [inline src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:136]: please handle disposable.

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts (hunk 2 above)
Comment (alanrenmsft) [reply src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:281]: this should be a warning or error instead of info, right?

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+
+		context.executionPlanDiagram.clearExpensiveOperatorHighlighting();
+		let result = context.executionPlanDiagram.highlightExpensiveOperator(expensiveOperationDelegate);
+		if (!result) {
+			context.notificationService.info(localize('invalidPropertyExecutionPlanMetric', 'Unable to locate a node using the specified metric.'));
+		}
+	}
+}
+
+export class TurnOffExpensiveOperationAction extends Action {
```
Comment (alanrenmsft) [inline src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:286]: please update the name to reflect the actual purpose: `TurnOffExpensiveHighlightingOperationAction`

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+	}
+
+	public override async run(context: FindExpensiveOperationWidget): Promise<void> {
+		context.executionPlanDiagram.clearExpensiveOperatorHighlighting();
+	}
+}
+
+export class CancelExpensiveOperationAction extends Action {
+	public static ID = 'qp.cancelExpensiveOperationAction';
+	public static LABEL = localize('cancelExpensiveOperationAction', 'Close (Escape)');
```
Comment (alanrenmsft) [inline src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:301]: same here, don't include the shortcut key in the text directly.

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+
+				this.showStoreDefaultMetricPrompt();
+			}
+			else if (ev.key === 'Escape') {
+				this.widgetController.removeWidget(self);
+			}
+		};
+
+		const applyButton = new Button(this.container, {
+			title: localize('findExpensiveOperationButtonTitle', 'Find Expensive Operation (Enter)')
```
[the remark this answers is not in the document]
Comment (lewis-sanchez) [reply src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:153]: I saw it being used in the `customZoomWidget`, and used it in the expensive node widget, but I will remove them both since I'm also not seeing it being used in other areas in the product.

File: src/sql/azdata.proposed.d.ts (hunk 3 above)
Comment (lewis-sanchez) [reply src/sql/azdata.proposed.d.ts]: Just a question, but does `Dictionary` not transfer well over RPC because it collides with C#'s Dictionary type, or is it just that Dictionary's don't serialize well?

File: src/sql/azdata.proposed.d.ts (hunk 6 above)
Comment (alanrenmsft) [reply src/sql/azdata.proposed.d.ts]: comment not resolved.

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+			}
+			else if (ev.key === 'Escape') {
+				this.widgetController.removeWidget(self);
+			}
+		};
+
+		const applyButton = new Button(this.container, {
+			title: localize('findExpensiveOperationButtonTitle', 'Find Expensive Operation')
+		});
+		applyButton.setWidth('60px');
```
Comment (alanrenmsft) [inline src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:155]: why are we setting the width explicitly? we should let it resize automatically because the word `resize` might be longer than 60px in a different language.

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+		this._register(findExpensiveOperationAction);
+
+		const clearExpensiveOperationAction = new TurnOffExpensiveHighlightingOperationAction();
+		this._register(clearExpensiveOperationAction);
+
+		const cancelExpensiveOperationAction = new CancelExpensiveOperationAction();
+		this._register(cancelExpensiveOperationAction);
+
+		const self = this;
+		this.operationNameSelectBox.selectElem.onkeydown = async (ev) => {
```
Comment (alanrenmsft) [inline src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:136]: for Enter key press, you shouldn't need to handle it, user is expected to use the apply button to apply changes. for escape key press, it is wrong to do it this way, the correct way is:
1. for mouse user, there should be a close button.
2. for keyboard user, the Escape key handling should happen at the pop out level, with your implementation, if the focus is on the Apply button, and user press Escape, the popup won't be closed.

File: src/sql/azdata.proposed.d.ts [hunk 13]
```diff
@@ -1338,9 +1338,32 @@ declare module 'azdata' {
...
+
+		export interface CostMetrics {
+			/**
+			 * CPU Time taken by the node operation in milliseconds
+			 */
+			elapsedCpuTimeInMs: number | undefined;
+			/**
+			 * Estimate number of rows for all executions.
+			 */
+			estimateRowsForAllExecutions: string | undefined;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts:1354]: this should be number

File: src/sql/azdata.proposed.d.ts [hunk 14]
```diff
@@ -1338,9 +1338,32 @@ declare module 'azdata' {
...
+			/**
+			 * Estimated number of rows read.
+			 */
+			estimatedRowsRead: string | undefined;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts:1358]: number

File: src/sql/azdata.proposed.d.ts [hunk 15]
```diff
@@ -1338,9 +1338,32 @@ declare module 'azdata' {
...
+			/**
+			 * The actual total number of rows.
+			 */
+			actualRows: string | undefined;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts:1362]: number

File: src/sql/azdata.proposed.d.ts [hunk 16]
```diff
@@ -1338,9 +1338,32 @@ declare module 'azdata' {
...
+			/**
+			 * The actual number of rows read.
+			 */
+			actualRowsRead: string | undefined;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts:1366]: number

File: src/sql/azdata.proposed.d.ts (hunk 6 above)
Comment (lewis-sanchez) [reply src/sql/azdata.proposed.d.ts]: Removed `Dictionary` and replaced with `CostMetrics` interface

File: src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts (hunk 11 above)
Comment (lewis-sanchez) [reply src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts:139]: I had to set the width explicitly because the widget select box looks like this if I don't:
image

With the width of 60px the select box looks like this:
image

Comment (lewis-sanchez) [reply src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts]: If a different language uses a longer string then the select box will look like this: 
image

Comment (alanrenmsft) [reply src/sql/workbench/contrib/executionPlan/browser/widgets/findExpensiveNodeWidget.ts]: there must be something else that is causing the size of the button to become that wide, check with developer tools, i suspect it is the width 100% plus the flex settings. by default, the button width is just the text plus some paddings.

File: src/sql/azdata.proposed.d.ts [hunk 17]
```diff
@@ -1333,6 +1337,33 @@ declare module 'azdata' {
...
+			 * Cost metrics for the node
+			 */
+			costMetrics: CostMetrics;
+		}
+
+		export interface CostMetrics {
+			/**
+			 * CPU Time taken by the node operation in milliseconds
+			 */
+			elapsedCpuTimeInMs: number | undefined;
```
Comment (aasimkhan30) [inline src/sql/azdata.proposed.d.ts:1350]: Can you please make this an array. This is not generic enough to be used in other databases.

File: src/sql/workbench/contrib/executionPlan/browser/azdataGraphView.ts [hunk 18]
```diff
@@ -292,37 +299,37 @@ export class AzdataGraphView {
...
+		diagramNode.relativeCost = node.relativeCost;
+		diagramNode.elapsedTimeInMs = node.elapsedTimeInMs;
 
-		if (node.children) {
-			diagramNode.children = [];
-			for (let i = 0; i < node.children.length; ++i) {
-				diagramNode.children.push(this.populate(node.children[i]));
-			}
-		}
+		diagramNode.costMetrics = {
```
Comment (aasimkhan30) [inline src/sql/workbench/contrib/executionPlan/browser/azdataGraphView.ts:325]: These things should become very easy if we make this array.

File: src/sql/workbench/contrib/executionPlan/browser/azdataGraphView.ts [hunk 19]
```diff
@@ -526,6 +584,17 @@ export enum SearchType {
 	LesserThanEqualTo,
 	LesserAndGreaterThan
 }
+
+export enum ExpensiveMetricType {
```
Comment (aasimkhan30) [inline src/sql/workbench/contrib/executionPlan/browser/azdataGraphView.ts:588]: It will be nice if the provider gives this from the backend.

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanView.ts [hunk 20]
```diff
@@ -26,7 +26,21 @@ import { Progress } from 'vs/platform/progress/common/progress';
 import { CancellationToken } from 'vs/base/common/cancellation';
 import { Action, Separator } from 'vs/base/common/actions';
 import { localize } from 'vs/nls';
-import { customZoomIconClassNames, disableTooltipIconClassName, enableTooltipIconClassName, executionPlanCompareIconClassName, executionPlanTopOperations, openPlanFileIconClassNames, openPropertiesIconClassNames, openQueryIconClassNames, savePlanIconClassNames, searchIconClassNames, zoomInIconClassNames, zoomOutIconClassNames, zoomToFitIconClassNames } from 'sql/workbench/contrib/executionPlan/browser/constants';
+import {
```
Comment (aasimkhan30) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanView.ts:29]: Lets just make this 

```
import * as constants from 'sql/workbench/contrib/executionPlan/browser/constants'
```

File: src/sql/workbench/contrib/executionPlan/browser/azdataGraphView.ts (hunk 18 above)
Comment (lewis-sanchez) [reply src/sql/workbench/contrib/executionPlan/browser/azdataGraphView.ts:325]: Would the array contain objects that resemble the following?
```
[
   { name: "metric-name", displayValue: 'display-value'}, ...
]

File: src/sql/workbench/contrib/query/browser/query.contribution.ts [hunk 21]
```diff
@@ -450,6 +450,21 @@ const queryEditorConfiguration: IConfigurationNode = {
...
+			'enumDescriptions': [
+				localize('queryEditor.executionPlan.expensiveOperationMetric.off', 'Expensive operation highlighting will be turned off for execution plans.'),
+				localize('queryEditor.executionPlan.expensiveOperationMetric.actualElapsedTime', 'Highlights the execution plan operation that took the most time.'),
+				localize('queryEditor.executionPlan.expensiveOperationMetric.actualElapsedCpuTime', 'Highlights the execution plan operation that used the most CPU time.'),
+				localize('queryEditor.executionPlan.cost', 'Highlights the execution plan operation with the highest cost.'),
+				localize('queryEditor.executionPlan.subtreeCost', 'Highlights the execution plan operation with the highest subtree cost.'),
+				localize('queryEditor.executionPlan.actualNumberOfRowsForAllExecutions', 'Highlights the execution plan operation with the greatest actual number of rows for all executions.'),
+				localize('queryEditor.executionPlan.numberOfRowsRead', 'Highlights the execution plan operation with the greatest number of rows read.')
+			],
+			'default': 'cost',
```
Comment (aasimkhan30) [inline src/sql/workbench/contrib/query/browser/query.contribution.ts:466]: Shouldn't this information come from mssql? Since this is sql server specific cost metric.
